pkg/distributor/distributor.go

```diff
@@ -633,7 +638,46 @@ func (d *Distributor) HealthyInstancesCount() int {
 	return int(d.healthyInstancesCount.Load())
 }
 
-func extractSampleSeries(req *distributormodel.PushRequest) []*distributormodel.ProfileSeries {
+type groupsWithFingerprints struct {
+	labelSets    []phlaremodel.Labels
+	groups       []pprof.SampleGroup
+	fingerprints map[uint64][]int
+}
+
+func newGroupsWithFingerprints(n int) *groupsWithFingerprints {
+	return &groupsWithFingerprints{
+		labelSets:    make([]phlaremodel.Labels, 0, n),
+		groups:       make([]pprof.SampleGroup, 0, n),
+		fingerprints: make(map[uint64][]int),
+	}
+}
+
+func (g *groupsWithFingerprints) add(lbls phlaremodel.Labels, group pprof.SampleGroup) {
+	fp := lbls.Hash()
+	idxs, ok := g.fingerprints[fp]
+	if ok {
+		// fingerprint matches, check if the labels arethe same
+		for _, idx := range idxs {
+			if phlaremodel.CompareLabelPairs(g.labelSets[idx], lbls) == 0 {
+				// append samples to the group
+				g.groups[idx].Samples = append(g.groups[idx].Samples, group.Samples...)
+				return
+			}
+		}
+	}
+
+	// add the labels to the list
+	g.fingerprints[fp] = append(g.fingerprints[fp], len(g.labelSets))
+	g.labelSets = append(g.labelSets, lbls)
+	g.groups = append(g.groups, group)
+
+}
+
+func extractSampleSeries(req *distributormodel.PushRequest, relabelRules []*relabel.Config) (result []*distributormodel.ProfileSeries, bytesRelabelDropped, profilesRelabelDropped float64) {
+	var (
+		lblbuilder = phlaremodel.NewLabelsBuilder(phlaremodel.EmptyLabels())
```

Comment:
I recently learned the `LabelBuilder` does not always guarantee uniqueness the hard way: https://github.com/grafana/pyroscope/pull/3356#discussion_r1639478243. Its performance also raises some concerns. It might be absolutely irrelevant, mentioning this just in case.